### PR TITLE
Open Graph: better video tags for Video posts including VideoPress videos

### LINKED ISSUE #1134: wpvideo shortcode: Undefined index when VideoPress inactive
When using a `[wpvideo xyz]` shortcode on a site when VideoPress is inactive, the following errors are thrown while putting together the OG tags:

Notice: Undefined index: id in /home/kraftbj/public_html/wp-content/plugins/jetpack/class.media-summary.php on line 63

Notice: Undefined index: id in /home/kraftbj/public_html/wp-content/plugins/jetpack/class.media-summary.php on line 64

The shortcode is still registered and renders correctly, besides the OG tag creation.

### LINKED ISSUE #2950: Open Graph: Stop using Flash for og:video
Right now, we have Flash hardcoded: https://github.com/Automattic/jetpack/blob/8a30daea4c9caea855c24b0ea7e9758d35e01e32/modules/publicize/enhanced-open-graph.php#L89

Per Otto, FB accepts mp4, which we should try to use, at least for VideoPress.

Would fix: https://meta.trac.wordpress.org/ticket/1375

### PR DESCRIPTION
**Pull request by jeherve**
Fixes #2950 
Fixes #1134
Replaces #2980

Replace the Flash `og:video` tags by the mp4 video link.

Also moves utility functions outside of the VideoPress module. Since the functions are used by more and more elements, even when the VideoPress module isn't active, it's best to make them accessible all the time.

### CODE REVIEW

**Comment by jeherve** [comment]
cc @dbtlr since you worked on it for WordPress.com.

Thanks!

File: _inc/lib/class.media-summary.php [hunk 1]
```diff
@@ -63,11 +63,34 @@ static function get( $post_id, $blog_id = 0, $args = array() ) {
 				switch ( $type ) {
 					case 'wpvideo':
 						if ( 0 == $return['count']['video'] ) {
-							$return['type'] = 'video';
-							$return['video'] = esc_url_raw( 'http://s0.videopress.com/player.swf?guid=' . $extract['shortcode']['wpvideo']['id'][0] . '&isDynamicSeeking=true' );
-							$return['image'] = self::get_video_poster( 'videopress', $extract['shortcode']['wpvideo']['id'][0] );
-							$return['secure']['video'] = preg_replace( '@http://[^\.]+.videopress.com/@', 'https://v0.wordpress.com/', $return['video'] );
-							$return['secure']['image'] = str_replace( 'http://videos.videopress.com', 'https://videos.files.wordpress.com', $return['image'] );
+							$guid = $extract['shortcode']['wpvideo']['id'][0];
+							$video_info = videopress_get_video_details( $guid );
```
Comment (dbtlr) [inline _inc/lib/class.media-summary.php]: I'm concerned about the implications of calling to this on every page load. We should definitely at least do some sort of caching routine. Though, long term, it as we start to merge more and more VideoPress things into the local library, we may find that we're able to do this call locally and rely on Jetpack sync in order to keep that data up to date.

File: modules/videopress/utility-functions.php [hunk 2]
```diff
@@ -1,99 +1,6 @@
 <?php
-
 /**
- * Validate user-supplied guid values against expected inputs
+ * Deprecated. Functions moved to functions.compat.php
```
Comment (georgestephanis) [inline modules/videopress/utility-functions.php]: Not deprecated.  New code will be coming in via #3838

File: _inc/lib/class.media-summary.php (hunk 1 above)
Comment (jeherve) [inline _inc/lib/class.media-summary.php]: Good call. I added some caching in 29a4ba25c54436900b60e7ef6f11fa4282416385

File: modules/videopress/utility-functions.php (hunk 2 above)
Comment (jeherve) [inline modules/videopress/utility-functions.php]: Oops, my bad. I removed "deprecated" in d7850ac60e9306126eb5f2aa2a97b0bbde247a0b

Also worth noting that I added some caching to the `videopress_get_video_details` function in 29a4ba25c54436900b60e7ef6f11fa4282416385, as per https://github.com/Automattic/jetpack/pull/3853#discussion_r64011347

**Comment by jeherve** [comment]
@georgestephanis Could I have your opinion on that PR? We discussed a bit about this at the meetup, but I don't remember what we agreed on.

Right now, all VideoPress utility functions are stored in `utility-functions.php`, and we call that file from the Shortcodes module as well as the VideoPress module. However, we'll need to access some of these utility functions outside of these 2 modules. In this PR for example, we need them when Sharing or Publicize is active.

I started moving these functions to `functions.compat.php`, but I seem to recall you suggesting another location for them.

How do you think I should proceed with this?

**Comment by dbtlr** [comment]
@jeherve My recollection was that it made sense to keep the `utility-functions.php` file, but to still move the functions you were moving out of it.

On a more future note, I suggest we leave it the way it is now, as I plan, after this next set of releases to do a full overhaul of the VideoPress module to clean up old code and to get things better organized, as I've had a hell of a time work with this and making changes because of how mixed together everything is. I'd prefer to have a refactor like that be stand alone from any feature releases, so we can isolate testing properly.

**Comment by jeherve** [comment]
Sounds good. At least this can wait until 4.1 is out the door, things should be a lot clearer then.

**Comment by samhotchkiss** [comment]
#5457

**Comment by dbtlr** [comment]
@jeherve: Where are we with getting this merged into 4.4?

**Comment by jeherve** [comment]
@dbtlr I rebased and reverted my changes to where the functions were located. It's ready for testing!

File: _inc/lib/class.media-summary.php [hunk 3]
```diff
@@ -63,11 +63,34 @@ static function get( $post_id, $blog_id = 0, $args = array() ) {
 				switch ( $type ) {
 					case 'wpvideo':
```
Comment (dbtlr) [inline _inc/lib/class.media-summary.php]: We should add in a fall-through for `case 'videopress':` as well, so that it shares this logic.

**Comment by dbtlr** [comment]
@jeherve I've tested and can confirm this works. I also made some changes to allow this to work with the `videopress` shortcode just like it works for the `wpvideo` shortcode.

File: modules/shortcodes.php [hunk 4]
```diff
@@ -134,4 +134,13 @@ function jetpack_shortcode_get_wpvideo_id( $atts ) {
 	}
 }
 
+if ( ! function_exists( 'jetpack_shortcode_get_videopress_id' ) ) {
+	function jetpack_shortcode_get_videopress_id( $atts ) {
+		if ( isset( $atts[ 0 ] ) )
```
Comment (dereksmart) [inline modules/shortcodes.php]: Can we please use brackets here?

Comment (jeherve) [inline modules/shortcodes.php]: Done in 8c70190f2e4d03eee44c1773651da36fe0e3c4bd, along with quick changes for coding standards in the whole file.

**Pull request merged**